cpp/src/copying/slice.cu

```diff
@@ -0,0 +1,296 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "types.hpp"
+#include "utilities/type_dispatcher.hpp"
+#include "utilities/error_utils.hpp"
+#include "utilities/cuda_utils.hpp"
+#include "rmm/thrust_rmm_allocator.h"
+#include <nvstrings/NVCategory.h>
+#include <bitmask/bit_mask.cuh> 
+
+namespace cudf {
+
+namespace {
+
+using bit_mask_t = bit_mask::bit_mask_t; 
+
+
+/**
+ * @brief Improve the readability of the source code.
+ * Parameter for the CUDA kernel.
+ */
+constexpr std::size_t NO_DYNAMIC_MEMORY = 0;
+
+
+template <typename ColumnType>
+__global__
+void slice_data_kernel(ColumnType*           output_data,
+                       ColumnType const*     input_data,
+                       gdf_index_type const* indices,
+                       gdf_index_type const  indices_position) {
+  
+  gdf_index_type input_offset = indices[indices_position*2];    /**< The start index position of the input data. */
+  gdf_size_type row_size = indices[indices_position*2 + 1] - input_offset; 
+
+  // Calculate kernel parameters
+  gdf_size_type row_index = threadIdx.x + blockIdx.x * blockDim.x;
+  gdf_size_type row_step = blockDim.x * gridDim.x;
+
+  // Perform the copying operation
+  while (row_index < row_size) {
+    output_data[row_index] = input_data[input_offset + row_index];
+    row_index += row_step;
+  }
+}
+
+/** @brief This function copies a slice of a bitmask.
+ * 
+ * If the slice is from element 10 to element 40, element 10 corresponds to bit 3 of the second byte, 
+ * that bit needs to become bit 0. So we are reading two adjacent blocks and bitshifting them together,
+ * to then write one block. We also take care that if the last bits of a bit_mask_t block don't 
+ * correspond to this slice, then we to apply a mask to clear those bits.
+*/
+__global__
+void slice_bitmask_kernel(bit_mask_t*           output_bitmask,
+                          gdf_size_type*        output_null_count,
+                          bit_mask_t const*     input_bitmask,
+                          gdf_size_type const   input_size,
+                          gdf_index_type const* indices,
+                          gdf_size_type const   indices_size,
+                          gdf_index_type const  indices_position) {
+  // Obtain the indices for copying
+  gdf_index_type input_index_begin = indices[indices_position * 2];
+  gdf_index_type input_index_end = indices[indices_position * 2 + 1];
+
+  gdf_index_type input_offset = input_index_begin / bit_mask::bits_per_element;
```

Comment:
I think this can be replaced with https://github.com/rapidsai/cudf/blob/branch-0.8/cpp/src/utilities/bit_util.cuh#L59

Reply:
done